CUDADataFormats/SiPixelCluster/interface/SiPixelClustersCUDA.h

```diff
@@ -5,16 +5,34 @@
 #include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/cudaCompat.h"
 
+#include "DataFormats/SoATemplate/interface/SoALayout.h"
+#include "CUDADataFormats/Common/interface/PortableDeviceCollection.h"
+
 #include <cuda_runtime.h>
 
-class SiPixelClustersCUDA {
+GENERATE_SOA_LAYOUT(SiPixelClustersCUDALayout,
+                    SOA_COLUMN(uint32_t, moduleStart),
+                    SOA_COLUMN(uint32_t, clusInModule),
+                    SOA_COLUMN(uint32_t, moduleId),
+                    SOA_COLUMN(uint32_t, clusModuleStart))
+
+using SiPixelClustersCUDASoA = SiPixelClustersCUDALayout<>;
+using SiPixelClustersCUDASOAView = SiPixelClustersCUDALayout<>::View;
+using SiPixelClustersCUDASOAConstView = SiPixelClustersCUDALayout<>::ConstView;
+
+// TODO: The class is created via inheritance of the PortableDeviceCollection.
+// This is generally discouraged, and should be done via composition, i.e.,
+// by adding a public class attribute like:
+// cms::cuda::Portabledevicecollection<SiPixelClustersCUDALayout<>> collection;
+// See: https://github.com/cms-sw/cmssw/pull/40465#discussion_r1067364306
+class SiPixelClustersCUDA : public cms::cuda::PortableDeviceCollection<SiPixelClustersCUDALayout<>> {
```

Comment:
I have a different question, though: there is no version that inherits from `PortableHostCollection`.
Is that because (in the CUDA code) we never copied this collection as-is back to the host ?

Reply:
Yes, we go back directly to legacy formats through the hits.